### ISSUE TITLE
Zed is continually writing to disk while idling

### ISSUE THREAD

**Issue by marvin-j97** [zed-industries/zed#16472]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Zed is continually writing to disk (about 200KB per second) even while idling, in any project, even empty projects.

See the gif for comparison with VS Code, which shows it's writing 0 while idling (right most column).

### Environment

Zed: v0.148.1 (Zed)
OS: Linux X11 ubuntu 22.04
Memory: 31.2 GiB
Architecture: x86_64
GPU: NVIDIA GeForce RTX 2080 Ti || NVIDIA || 545.29.06

### If applicable, add mockups / screenshots to help explain present your vision of the feature

![Peek 2024-08-19 17-38](https://github.com/user-attachments/assets/8cd0f3dd-c021-4fd5-9a47-42d840c4c092)

### If applicable, attach your Zed.log file to this issue.

Logs are not showing anything after startup

**Comment by rohitshriwas** [comment]
I can confirm this behaviour on Zed v0.152.4 on Linux (Flatpak). Is this bad for SSD?

![image](https://github.com/user-attachments/assets/2bab3e4a-b134-46b8-8b76-2df22a788801)

**Comment by pepyakin** [comment]
I was flying on a plane and realized that my battery is drained if not faster but definitely on par with vscode. I decided to take a look and realized the it's the excessive disk writes.

The writes are coming from sqlite. They are mostly invoked WorkspaceDb. 

If you look where `Workspace::serialize_workspace` is invoked it's basically coming from `cx.observe` on dock views. What's funny is that it is triggered from `-[NSWindow _handleMouseDraggedEvent:]`. So each time you drag a mouse it ultimatelly triggers a sqlite write. To add salt to the wound, all the sqlite statements are NOT prepared.

**Comment by kingsidelee** [comment]
you can workaround this by open up zed using envvar `ZED_STATELESS=1`, it will force the program to use a in-memory sqlite db.
https://github.com/zed-industries/zed/blob/e4eef725de16025f8133013a5e8a1f53c29b7c79/crates/db/src/db.rs#L49-L51
this won't solve the performance issue, it will still waste some cpu cycles to write the in-memory db.

**Comment by SomeoneToIgnore** [comment]
Thank you for the great `observe` find, I've merged https://github.com/zed-industries/zed/pull/22730 to fix that.
It gets released tomorrow, and, on my tests, removes most of the "extra" writes to rerialize the workspace — would be really interested to see if anything's left to fix after your testing.

**Comment by pepyakin** [comment]
@SomeoneToIgnore can confirm it makes a significant improvement!

That being said, there is some work to do. For example, when you click on an item in the project panel, it still does like 5 calls to serialize_workspace. Can we debounce it by increasing the delay up to 1s?

**Comment by SomeoneToIgnore** [comment]
Yes, this is closer to a stub than a full bulletproof fix, alas.

I wanted to bump the serialization debounce, but https://github.com/zed-industries/zed/pull/22730#discussion_r1904757829 was voiced against that so it's not the way we're following it seems.

Still, even with the current value, for a patch
```diff
diff --git a/crates/workspace/src/workspace.rs b/crates/workspace/src/workspace.rs
index e9676dd4a0..92b10b4d61 100644
--- a/crates/workspace/src/workspace.rs
+++ b/crates/workspace/src/workspace.rs
@@ -4130,6 +4130,7 @@ impl Workspace {
         let Some(database_id) = self.database_id() else {
             return Task::ready(());
         };
+        dbg!("serializing the workspace");

         fn serialize_pane_handle(pane_handle: &View<Pane>, cx: &WindowContext) -> SerializedPane {
             let (items, active, pinned_count) = {
```

that's in the serialization start, I see only one actual serialization when clicking the project panel items:


https://github.com/user-attachments/assets/a93288f8-7769-4c4c-8a3c-4b18e356e7f7


Which seems to work as expected, given that workspace has to remember which item to show on restore.
Do you see a different picture with this patch or is it the DB writes themselves that you see different? 
Curious what's not matching in our set-ups.

I have tested all the panels before merging this, and all of them seem to work in a similar way (1 or 0 serialization attempts) except the outline panel which is changing the active item and cause more serialization than needed.
That bit I plan to patch later, but otherwise not noticing any related issues anymore.

**Comment by pepyakin** [comment]
Oh, sorry, I am retarded. The last few times I tested I was setting a breakpoint on `serialize_workspace`. Then it looks great! 

I am wondering though what would the bulletproof fix look like. Just curiousity, feel free to ignore.

The comment re debounce makes sense, although I am not sure if I am 100% on board with it. It's not like Zed crashes that often, nor can I shutdown Zed in 300ms, nor I care saving the last 300ms of progress. cc @mikayla-maki

**Comment by SomeoneToIgnore** [comment]
I do not have a very firm idea, but to handwave a bit around the good fix:

* the serialization should compare previous serialized state with the new one and only write into DB if anything's changed

* the serialization should react on event rather than a `.serialize_workspace` method call called somewhere outside of the workspace.rs

* there has to be a basic test of user interactions that counts the serialization attempts (the hardest part, as our tests do not really construct the entire Zed as it's constructed during startup + we have no e2e tests like that yet)

**Comment by mikayla-maki** [comment]
I'm not observing this specific issue with workspace serialization, and since @SomeoneToIgnore fixed some major cases of over writing, I'm going to close this issue for now. Let us know if it comes back! :)

**Comment by marvin-j97** [comment]
> I'm not observing this specific issue with workspace serialization, and since [@SomeoneToIgnore](https://github.com/SomeoneToIgnore) fixed some major cases of over writing, I'm going to close this issue for now. Let us know if it comes back! :)

This is still happening in 0.175.5

**Comment by SomeoneToIgnore** [comment]
How can I reproduce this?
Anything specific you are doing?

Have rechecked the workspace serialization case and that only happens when expected now — e.g. on panel toggling, nothing changes the value when just hovering the mouse or doing nothing.

<img width="876" alt="Image" src="https://github.com/user-attachments/assets/dfef1cd2-1917-445c-b86d-78725c0b7c1c" />

**Comment by SomeoneToIgnore** [comment]
Sad to see this, but 80kb/s hints me Zed's more likely writing something else, not DB this time (as relatively small for a workspace serialized).

@marvin-j97 luckily, you're on Linux, so we can investigate it better: can you use `lsof` or `strace` or similar to check what exactly does Zed's process access and writes to?

Could Zed be writing something into its log file instead of the DB for you? I recall X11 had issues with that but failed to find an issue right now.

**Comment by cels** [comment]
I also have this issue.
- up-to-date Arch Linux
- zed 0.181.6
- constant IO of ~80kb/s

I did some digging as mentioned by @SomeoneToIgnore:

First narrowed down files with current timestamp with help of `lsof` which were these:
`~/.local/share/zed/db/0-stable/db.sqlite`
`~/.local/share/zed/db/0-stable/db.sqlite-shm`
`~/.local/share/zed/db/0-stable/db.sqlite-wal`

Checked files with `watch -d -n 1 ls -la --time-style=full-iso ~/.local/share/zed/db/0-stable/db.sqlite*` to see which one changes constantly, result:

![Image](https://github.com/user-attachments/assets/3b7d5f66-b17d-48d7-9705-4eb537ca8f44)

As highlighted in the ouput, the culprit seems to be `/.local/share/zed/db/0-stable/db.sqlite-wal`

@SomeoneToIgnore is there anything more I can do to help with this?

**Comment by SomeoneToIgnore** [comment]
Thank you for checking this.

I cannot think of anything simple, but one can build and run Zed locally with extra logging added.

We can check all `persistence.rs` modules and their methods where we write anything into the DB:

<img width="552" alt="Image" src="https://github.com/user-attachments/assets/b68e6cda-9c44-451a-8164-aa8b4c18a08c" />

The most suspicious are `workspace` and `editor` ones: the former caused troubles in this issue and might have been fixed (as it was [last time](https://github.com/zed-industries/zed/issues/16472#issuecomment-2692794587) I checked this), the latter is quite fundamental and may cause extra writes now when I revisit and think of it.

Unfortunately, we do not seem to have any logs there, but now we have [`zlog`](https://github.com/zed-industries/zed/pull/28544) and it might be a good PR to add some `log::debug!` ones there to ease the future investigations.

Even without it, we could just put a bunch of `dbg!` inside every `INSERT`/`UPDATE` DB-related methods to see what is being called that much.
Or just comment out bodies of the methods to check if things are improving.


I will try to repro the issue more, thank you for the command example, will see if I can reproduce anything related myself.

**Comment by cels** [comment]
Thanks!

Unfortunately I have no clue about rust, otherwise I could try something myself.

@SomeoneToIgnore please let me know when there is a branch to test or so, maybe I can at least get building Zed running.

**Comment by SomeoneToIgnore** [comment]
Thank you again for the checks, I was under the impression issues are mainly related to the workspace serialization but indeed it's not correct.

https://github.com/zed-industries/zed/pull/29417 should hopefully fix your case or, at worst, provide ways to detect more issues without the recompilation.

**Comment by cels** [comment]
Oh, you already fixed it :-)

Thanks!
I'll keep you updated once the update reaches me.
